drivers/sensor/veml7700/veml7700.c

```diff
@@ -0,0 +1,555 @@
+ * Those two are used for the high and low threshold
+ * interrupt flags in the ALS_INT register.
+ */
+#define VEML7700_ALS_INT_MASK (~BIT_MASK(14))
+
+/*
+ * 16-bit command register addresses
+ */
+#define VEML7700_CMDCODE_ALS_CONF 0x00
+#define VEML7700_CMDCODE_ALS_WH	  0x01
+#define VEML7700_CMDCODE_ALS_WL	  0x02
+#define VEML7700_CMDCODE_PSM	  0x03
+#define VEML7700_CMDCODE_ALS	  0x04
+#define VEML7700_CMDCODE_WHITE	  0x05
+#define VEML7700_CMDCODE_ALS_INT  0x06
+
+/*
+ * Devicetree psm-mode property value for "PSM disabled"
+ */
+#define VEML7700_PSM_DISABLED 0x00
+
+/*
+ * ALS integration time setting values.
+ *
+ * The enumerators of <tt>enum veml7700_als_it</tt> provide
+ * indices into this array to get the related value for the
+ * ALS_IT configuration bits.
+ */
+static const uint8_t veml7700_it_values[VEML7700_ALS_IT_ELEM_COUNT] = {
+	0x0C, /* 25  - 0b1100 */
+	0x08, /* 50  - 0b1000 */
+	0x00, /* 100 - 0b0000 */
+	0x01, /* 200 - 0b0001 */
+	0x02, /* 400 - 0b0010 */
+	0x03, /* 800 - 0b0011 */
+};
+
+/*
+ * Resolution matrix for values to convert between data provided
+ * by the sensor ("counts") and lux.
+ *
+ * These values depend on the current gain and integration time settings.
+ * The enumerators of <tt>enum veml7700_als_gain</tt> and <tt>enum veml7700_als_it</tt>
+ * are used for indices into this matrix.
+ */
+static const float veml7700_resolution[VEML7700_ALS_GAIN_ELEM_COUNT][VEML7700_ALS_IT_ELEM_COUNT] = {
+	/* 25ms    50ms    100ms   200ms   400ms   800ms  Integration Time */
+	{0.2304, 0.1152, 0.0576, 0.0288, 0.0144, 0.0072}, /* Gain 1 */
+	{0.1152, 0.0576, 0.0288, 0.0144, 0.0072, 0.0036}, /* Gain 2 */
+	{1.8432, 0.9216, 0.4608, 0.2304, 0.1152, 0.0576}, /* Gain 1/8 */
+	{0.9216, 0.4608, 0.2304, 0.1152, 0.0576, 0.0288}, /* Gain 1/4 */
+};
+
+struct veml7700_config {
+	struct i2c_dt_spec bus;
+	uint8_t psm;
+};
+
+struct veml7700_data {
+	uint8_t shut_down;
+	enum veml7700_als_gain gain;
+	enum veml7700_als_it it;
+	enum veml7700_int_mode int_mode;
+	uint16_t thresh_high;
+	uint16_t thresh_low;
+	uint16_t als_counts;
+	uint32_t als_lux;
+	uint32_t int_flags;
+};
+
+/**
+ * @brief Waits for a specific amount of time which depends
+ * on the current integration time setting.
+ *
+ * According to datasheet for a measurement to complete one has
+ * to wait for at least the integration time. But tests showed
+ * that a much longer wait time is needed. Simply adding 50 or
+ * 100ms to the integration time is not enough so we doubled
+ * the integration time to get our wait time.
+ */
+static void veml7700_sleep_by_integration_time(const struct veml7700_data *data)
+{
+	switch (data->it) {
+	case VEML7700_ALS_IT_25:
+		k_msleep(50);
+		break;
+	case VEML7700_ALS_IT_50:
+		k_msleep(100);
+		break;
+	case VEML7700_ALS_IT_100:
+		k_msleep(200);
+		break;
+	case VEML7700_ALS_IT_200:
+		k_msleep(400);
+		break;
+	case VEML7700_ALS_IT_400:
+		k_msleep(800);
+		break;
+	case VEML7700_ALS_IT_800:
+		k_msleep(1600);
+		break;
```

Comment:
I hoped it would be clear by the function documentation. I'm gonna add a few more lines with additional explanations and a reference to the application notes for all the lore.